cloudinit/sources/DataSourceEc2.py

```diff
@@ -731,28 +735,97 @@ def convert_ec2_metadata_network_config(network_md, macs_to_nics=None,
     """
-    netcfg = {'version': 1, 'config': []}
+    netcfg = {'version': 2, 'ethernets': {}}
     if not macs_to_nics:
         macs_to_nics = net.get_interfaces_by_mac()
     macs_metadata = network_md['interfaces']['macs']
-    for mac, nic_name in macs_to_nics.items():
+
+    if not apply_network_config:
+        for mac, nic_name in macs_to_nics.items():
+            if nic_name == fallback_nic:
+                break
+        dev_config = {'dhcp4': True,
+                      'dhcp6': False,
+                      'match': {'macaddress': mac.lower()},
+                      'set-name': nic_name}
+        nic_metadata = macs_metadata.get(mac)
+        if nic_metadata.get('ipv6s'):  # Any IPv6 addresses configured
+            dev_config['dhcp6'] = True
+        netcfg['ethernets'][nic_name] = dev_config
+        return netcfg
+    # Apply network config for all nics and any secondary IPv4/v6 addresses
+    nic_idx = 1
+    for mac, nic_name in sorted(macs_to_nics.items()):
         nic_metadata = macs_metadata.get(mac)
         if not nic_metadata:
             continue  # Not a physical nic represented in metadata
-        nic_cfg = {'type': 'physical', 'name': nic_name, 'subnets': []}
-        nic_cfg['mac_address'] = mac
-        if (nic_name == fallback_nic or nic_metadata.get('public-ipv4s') or
-                nic_metadata.get('local-ipv4s')):
-            nic_cfg['subnets'].append({'type': 'dhcp4'})
-        if nic_metadata.get('ipv6s'):
-            nic_cfg['subnets'].append({'type': 'dhcp6'})
-        netcfg['config'].append(nic_cfg)
+        dhcp_override = {'route-metric': nic_idx * 100}
+        nic_idx += 1
+        dev_config = {'dhcp4': True, 'dhcp4-overrides': dhcp_override,
+                      'dhcp6': False,
+                      'match': {'macaddress': mac.lower()},
+                      'set-name': nic_name}
+        if nic_metadata.get('ipv6s'):  # Any IPv6 addresses configured
+            dev_config['dhcp6'] = True
+            dev_config['dhcp6-overrides'] = dhcp_override
+        dev_config['addresses'] = get_secondary_addresses(nic_metadata, mac)
+        if not dev_config['addresses']:
+            dev_config.pop('addresses')  # Since we found none configured
+        netcfg['ethernets'][nic_name] = dev_config
     return netcfg
 
 
+def get_secondary_addresses(nic_metadata, mac):
+    """Parse interface-specific nic metadata and return and any secondary IPs
+
+    :return: List of secondary IPv4 or IPv6 addresses to configure on the
+    interface
+    """
+    ipv4s = nic_metadata.get('local-ipv4s')
+    ipv6s = nic_metadata.get('ipv6s')
+    # In version < 2018-09-24 local_ipv4s or ipv6s is a str with one IP
+    secondary_ipv4 = bool(isinstance(ipv4s, list) and len(ipv4s) > 1)
+    secondary_ipv6 = bool(isinstance(ipv6s, list) and len(ipv6s) > 1)
+    if not any([secondary_ipv4, secondary_ipv6]):
+        return []
+    addresses = []
+    ip_metadata = {
```

Comment:
I find this implementation a little confusing. Partly, I think, it's because this _isn't_ metadata, it's logic encoded into a dict, which took me a couple of parses to figure out. Generally, I consider looping over a statically-defined dict to be an indication that we could be using first-class control flow instead.

In this case, we're using the dict to execute the same code twice with different parameters, which made me wonder if a function-based approach would work instead. This is a rough (totally untested) sketch of the sort of thing I'm thinking of:

```py
def _get_secondary_addresses(cidr_key: str, ips: List[str], prefix: str) -> List[str]:
    cidr = nic_metadata.get(cidr_key)
    if not cidr or len(cidr.split('/')) != 2:
        LOG.warning(...)
    else:
        prefix = cidr.split('/')[1]
    # We know we have > 1 ips for in metadata for this IP type
    return ['{ip}/{prefix}'.format(ip=ip, prefix=prefix) for ip in ips[1:]]

addresses = []
if bool(isinstance(ipv4s, list) and len(ipv4s) > 1):
    addresses.extend(_get_secondary_addresses('subnet-ipv4-cidr-block', ipv4s, '24'))
if bool(isinstance(ipv6s, list) and len(ipv6s) > 1):
    addresses.extend(_get_secondary_addresses('subnet-ipv6-cidr-block', ipv6s, '128'))

return sorted(addresses)
```

(Not saying you need to include the type annotations, just including them for additional clarity here.)

What do you think?

Reply:
yes this makes sense and makes things cleaner and easier to read. Thanks for that.